Добавить бейджи для линтеров в README.md (#33)

birthdaysgift opened the pull request:
Мы хотим отобразить статусы прохождения нашего пайплайна для каждого линтера (и пайтеста) с помощью бейджей в README.md

По результатам ресёрча было найдено три возможных решения:
- [Github](https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/adding-a-workflow-status-badge) badges проблема этого подхода в том, что гитхаб обновляет бейджи слишком [долго](https://stackoverflow.com/questions/57719605/official-badge-for-github-actions#answer-57836888) (около 10 минут после пуша), а также он позволяет создавать бейдж только для целого воркфлоу, но все наши три линтера находятся в одном воркфлоу и при этом мы хотим отобразить разные бейджи для разных линтеров. Поэтому этот вариант нам не подходит

- [BYOB](https://github.com/marketplace/actions/bring-your-own-badge) action проблема этого подхода в том, что он использует отдельную ветку, чтобы хранить и обновлять данные для бейджей, и нам бы не хотелось иметь лишнюю ветку в графе гита не относящуюся к работе

- [Dynamic Badges](https://github.com/marketplace/actions/dynamic-badges) action обновляет бейджи за 5 минут. Использует приватный GitHub Gist для хранения данных для бейджей а также даёт определенную гибкость в настройке бейджей (например, позволяет установить диапазон цветов для бейджа) - так что этот вариант и будет нашим выбором

В рамках этой задачи необходимо:
_(большинство шагов взяты из [доки](https://github.com/marketplace/actions/dynamic-badges#configuration) Dynamic Badges)_

- создать новый приватный гист, назвать его, например, `flake8.json` и скопировать ID гиста (это длинная цифро-буквенная часть url, по которому находится гист)

- перейти в [github.com/settings/tokens](https://github.com/settings/tokens) и создать новый токен со скоупом прав для работы с гистами

- перейти на страницу Secrets в настройках репозитория (Settings > Secrets > Actions) и добавить этот токен как новый секрет

- добавить шаг создания гиста в воркфлоу:
```yaml
  - name: Create Awesome Badge
    uses: schneegans/dynamic-badges-action@v1.6.0
    with:
      auth: ${{ secrets.GIST_SECRET }}
      gistID: <gist-ID>
      filename: flake8.json
      label: Hello
      message: World
      color: orange
```

- добавить бейдж в README.md с помощью ссылки: ``` ![badge](https://img.shields.io/endpoint?url=https://gist.githubusercontent.com/<user>/<gist-ID>/raw/flake8.json) ```

- добавить вышеописанный шаг в джобу после каждого линтера и связать их через `needs: [ ... ]`

- настроить этот бейдж, чтобы он запускался всегда вне зависимости от того, отработала джоба с линтером или упала с помощью `if: ${{ always() }}`

- чтобы получить результат работы джобы линтера, необходимо использовать `needs.<job-id>.result в данном случае нас интересует только значение `success`

- т.к. синтаксис для воркфлоу не позволяет использовать тернарные операторы, то можно использовать следующий код, чтобы выбирать разный цвет бейджа в зависимости от состояния джобы с линтером: ```yaml color: ${{ needs.mypy.result == 'success' && '#238636' || '#6b2a2b' }} ``` этот код работает точно так же как это тернарное выражение: ```javascript needs.mypy.result == 'success' ? '#238636' : '#6b2a2b' ```

Rey-Lane reviewed (commented):
Цвета страшные, конечно, ахаха
А мы не можем их куда-то вынести, чтобы можно было одним разом поменять, если захочется?

birthdaysgift commented:
> А мы не можем их куда-то вынести, чтобы можно было одним разом поменять, если захочется?

Вынес в переменные окружения, которые можно настроить в репозитории гитхаба:
![image](https://github.com/week-password/wisher-backend/assets/35929293/2b09fac6-fa85-4940-95a9-2d81499fbec1)

Цвета пока что оставил те же, что и были. Но теперь если появится желание поиграться с цветами, можно будет это сделать не меняя код.

birthdaysgift committed:
Добавить бейджи для линтеров в README.md (#33)

Мы хотим отобразить статусы прохождения нашего пайплайна для каждого линтера (и пайтеста) с помощью бейджей в README.md

По результатам ресёрча было найдено три возможных решения:
- [Github](https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/adding-a-workflow-status-badge) badges
  проблема этого подхода в том, что гитхаб обновляет бейджи слишком [долго](https://stackoverflow.com/questions/57719605/official-badge-for-github-actions#answer-57836888) (около 10 минут после пуша), а также он позволяет создавать бейдж только для целого воркфлоу, но все наши три линтера находятся в одном воркфлоу и при этом мы хотим отобразить разные бейджи для разных линтеров. Поэтому этот вариант нам не подходит

- [BYOB](https://github.com/marketplace/actions/bring-your-own-badge) action
  проблема этого подхода в том, что он использует отдельную ветку, чтобы хранить и обновлять данные для бейджей, и нам бы не хотелось иметь лишнюю ветку в графе гита не относящуюся к работе

- [Dynamic Badges](https://github.com/marketplace/actions/dynamic-badges) action
  обновляет бейджи за 5 минут. Использует приватный GitHub Gist для хранения данных для бейджей а также даёт определенную гибкость в настройке бейджей (например, позволяет установить диапазон цветов для бейджа) - так что этот вариант и будет нашим выбором

В рамках этой задачи необходимо:
_(большинство шагов взяты из [доки](https://github.com/marketplace/actions/dynamic-badges#configuration) Dynamic Badges)_

- создать новый приватный гист, назвать его, например, `flake8.json` и скопировать ID гиста
  (это длинная цифро-буквенная часть url, по которому находится гист)

- перейти в [github.com/settings/tokens](https://github.com/settings/tokens) и создать новый токен со скоупом прав для работы с гистами

- перейти на страницу Secrets в настройках репозитория (Settings > Secrets > Actions) и добавить этот токен как новый секрет

- добавить шаг создания гиста в воркфлоу:
```yaml
  - name: Create Awesome Badge
    uses: schneegans/dynamic-badges-action@v1.6.0
    with:
      auth: ${{ secrets.GIST_SECRET }}
      gistID: <gist-ID>
      filename: flake8.json
      label: Hello
      message: World
      color: orange
```

- добавить бейдж в README.md с помощью ссылки:
  ```
  ![badge](https://img.shields.io/endpoint?url=https://gist.githubusercontent.com/<user>/<gist-ID>/raw/flake8.json)
  ```

- добавить вышеописанный шаг в джобу после каждого линтера и связать их через `needs: [ ... ]`

- настроить этот бейдж, чтобы он запускался всегда вне зависимости от того, отработала джоба с линтером или упала с помощью `if: ${{ always() }}`

- чтобы получить результат работы джобы линтера, необходимо использовать `needs.<job-id>.result
  в данном случае нас интересует только значение `success`

- т.к. синтаксис для воркфлоу не позволяет использовать тернарные операторы, то можно использовать следующий код, чтобы выбирать разный цвет бейджа в зависимости от состояния джобы с линтером:
  ```yaml
  color: ${{ needs.mypy.result == 'success' && '#238636' || '#6b2a2b' }}
  ```
  этот код работает точно так же как это тернарное выражение:
  ```javascript
  needs.mypy.result == 'success' ? '#238636' : '#6b2a2b'
  ```
